Lf 3334 investigate node 16 eol sept 2023

Duncan-Brain opened the pull request:
**Description**

Upgrading Node 16 to Node 18.16.1 LTS and dealing with the issues that follow from that.

Overview of what was done:
All
- package-lock.json rebuilt

Frontend
- Material UI was upgraded using the fairly thorough [Material UI Migration Guide](https://mui.com/material-ui/migration/migration-v4/) and pre-made codemods
  - Codemods were run
  - Guide was followed for non code mods elligible fixes
  - some minor style changes
  - Frontend was toured by identifying all custom components that use MUI and visiting some view  
- React-Icons was fixed with new import syntax
  - `import {VSCDialog} from react-icons/all` --> `import {VSCDialog} from react-icons/vsc`

Backend
- Objection JS also has a brief migration guide but we were pretty far behind [Objection Migration Guide](https://vincit.github.io/objection.js/release-notes/migration.html)
  - type: date --> format: date (see ajv formats) and https://github.com/Vincit/objection.js/issues/2142 there is now a strict adherence to jsonSchema.
  - null --> 'null' on types and vice versa for enums
  - Union types not allowed: type: ['string', 'number] --> type: {anyOf: {type:'string'}, ,{type:'number'}} https://github.com/fastify/fastify/issues/3984
  - Random but hard to find schema mismatches (ie required: true is not a thing)
  - Defer to supported validator Ajv formats for validating special formatting: https://github.com/Vincit/objection.js/issues/2146#issuecomment-971738698
- Ajv formats was added to give the functionality we thought we had, but was actually probably not enforced with the jsonSchema that preceded Objection V3
  - BaseModel updates to $parseJson to handle null, object and special dates to pass validation that are handled treated differently in a variety of places including $BeforeInsert, server.js, and Jest
- A few Jest test updates
  - quantity_kg --> quantity (but not quantiy_kg/unit)
  - Added seed for field_work_task_type
- Objection soft delete
  - Patch mergeContext with prototype assignment https://github.com/griffinpp/objection-soft-delete/issues/18

Other:
Why not use $BeforeValidate: https://github.com/Vincit/objection.js/issues/77
Why I did not use objection-js-soft-delete (fix does not work in jest): https://github.com/alex-w0/objection-js-soft-delete/issues/116
What made me try objection-js-soft-delete several times: https://github.com/griffinpp/objection-soft-delete/issues/18
Testing out of use routes like Pesticide with Postman requires Body data in delete request (bad pattern) to test anyway comment out rejectInBodyGetAndDelete in server.js .



Jira link: [LF-3334](https://lite-farm.atlassian.net/browse/LF-3334)

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain):
- Please run backend Jest Tests they should Pass (this is how I found many of the classes of type errors from ajv format and schema validation)
- Frontend basic functionality seems to all be working scrolling around and testing
- I would love it if someone could test cypress test to see if it picks anything up (@MwayaB are any tests working again?)
- Look for: 
    - Whitescreen crashes with material UI console statements on frontend
    - 400 bad requests on the backend likely due to a formatting error on a Model validation. Use .debug() on the model validation if error not visible

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files


[LF-3334]: https://lite-farm.atlassian.net/browse/LF-3334?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

iperdomo commented:
This branch should be updated with the latest `integration`. Note the conflicts on `packages/api/package-lock.json`, `packages/webapp/pnpm-lock.yaml`

iperdomo commented on packages/api/src/controllers/loginController.js:
```diff
@@ -131,8 +131,8 @@ const loginController = {
           await UserModel.transaction(async (trx) => {
-            await UserModel.query(trx).insert(newUser);
-            await ShowedSpotlightModel.query(trx).insert({ user_id });
+            await UserModel.query(trx).insert(newUser).debug();
```
Are this `.debug()` calls needed?
https://vincit.github.io/objection.js/api/query-builder/other-methods.html#debug

Duncan-Brain replied:
Thank you! missed that!

iperdomo reviewed (changes requested):
Great job! ... I haven't tested locally but I can imagine all the work going through all those little details.

One thing I noticed is that we didn't updated [prod.Dockerfile](https://github.com/LiteFarmOrg/LiteFarm/blob/c140dc32e0cba664437eb05aaeb9b083c51cfd92/packages/api/prod.Dockerfile#L1)

Duncan-Brain commented:
@iperdomo Thanks totally forgot those two items.

I did not touch storybook.Dockerfile .. which seems to need updating as well.

iperdomo commented:
@Duncan-Brain I was trying to produce a Docker image from `webapp` and got this error
```
#14 [build 7/7] RUN pnpm run build
#14 0.895 
#14 0.895 > litefarm-webapp@3.3.2 build /usr/src/app
#14 0.895 > tsc && NODE_OPTIONS=--max_old_space_size=3000 vite build
#14 0.895 
#14 9.574 src/components/Typography/index.tsx(197,13): error TS2322: Type 'true | ReactPortal | ReactChild | ReactFragment' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
#14 9.574   Type 'number' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
#14 9.623  ELIFECYCLE  Command failed with exit code 2.
#14 ERROR: process "/bin/sh -c pnpm run build" did not complete successfully: exit code: 1
------
 > [build 7/7] RUN pnpm run build:
0.895 
0.895 > litefarm-webapp@3.3.2 build /usr/src/app
0.895 > tsc && NODE_OPTIONS=--max_old_space_size=3000 vite build
0.895 
9.574 src/components/Typography/index.tsx(197,13): error TS2322: Type 'true | ReactPortal | ReactChild | ReactFragment' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
9.574   Type 'number' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
9.623  ELIFECYCLE  Command failed with exit code 2.
------
prod.Dockerfile:13
--------------------
  11 |     COPY ./shared/ /usr/src/shared/
  12 |     
  13 | >>> RUN pnpm run build
  14 |     
  15 |     FROM nginx:1.15
--------------------
ERROR: failed to solve: process "/bin/sh -c pnpm run build" did not complete successfully: exit code: 1
```
The interesting bit
```
9.574 src/components/Typography/index.tsx(197,13): error TS2322: Type 'true | ReactPortal | ReactChild | ReactFragment' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
9.574   Type 'number' is not assignable to type 'string | (string & ReactElement<any, string | JSXElementConstructor<any>>) | (string & Iterable<ReactNode>) | (string & ReactPortal) | undefined'.
```

Duncan-Brain commented:
@iperdomo I am not sure why that only just now presented itself to be honest. I fixed it though by resolving the type error in the Typography Tooltip component. That made me wonder if other utilities from our scripts might have issues and it seems like Storybook tests are no longer passing (not sure if it was this or pre-existing but hoping to determine tomorrow).

I also found that I forgot to update package-lock after merging integration (necessary because of the npm vulnerable packages). @SayakaOno there was a change to the overrides made in /webapp -- the vite override was causing my modules not to allow lazy loading. I deleted the overrides and re-ran pnpm audit fix and that specific override did not rejoin.

@kathyavini I added the node update to the deploy script thanks!

kathyavini reviewed (approved):
I've been using this branch exclusively for several days and it's great -- super stable!! I see one deprecation warning on the frontend which I think can be its own ticket?

There's a good chance we'll have to do a one-time `nvm install` on the beta export server droplet before we can `nvm use 18.16.1` (so no worries if it fails the first time). I just added updating the prod version of our export server to our release checklist -- hopefully our version of pm2 is compatible with Node 18.16.1 🤞 -- Is there anything else related to this update that should go on the release checklist?

And you really held firm on not updating the Docker Hub Image reference 😉 I'll open a PR once this one is merged, then.

Wonderful job on another massive undertaking 🙂

kathyavini reviewed (approved):
I'm not getting close to only 33 failing Storybook tests, but that's an issue for another day!

Duncan-Brain commented on packages/webapp/src/assets/theme/index.tsx:
```diff
@@ -1,6 +1,13 @@
-import { createTheme, ThemeProvider } from '@material-ui/core';
+import { createTheme, StyledEngineProvider } from '@mui/material';
+import { adaptV4Theme, ThemeProvider, Theme } from '@mui/material/styles';
```
definitely another time! If it ain't broke..

kathyavini replied:
Alright I'll make a ticket

SayakaOno commented:
I confirmed that tests for the unit component ran successfully for me, too. I'm not able to run tests in integration branch, so I can't compare the result, but I know all tests were not passing before, and I guess it should be good!
Here is the result:
<img width="335" alt="Screenshot 2023-06-28 at 2 22 43 AM" src="https://github.com/LiteFarmOrg/LiteFarm/assets/33141219/13a85e25-f6e5-48fe-9464-97e13f1bd6c5">
